### PR TITLE
ARROW-12661: [C++] Add ReaderOptions::skip_rows_after_names

### PR DESCRIPTION
**Pull request by n3world**
Add a new csv reader option which allows the reader to skip rows after reading the column names from the csv.

### CODE REVIEW

File: cpp/src/arrow/csv/reader_test.cc
Comment (westonpace) [inline cpp/src/arrow/csv/reader_test.cc:219]: Can you add a test case where `autogenerate_column_names` is set to `true` (i.e. there is no header in the CSV file)?

Comment (n3world) [reply cpp/src/arrow/csv/reader_test.cc:219]: Done

**Comment by westonpace** [comment]
Thanks for submitting these PRs!  I'm still brushing up on my parser knowledge but I'll take a look at the other one too.

This one looks pretty straightforward to me.  It's a bit unfortunate that we have two `skip_rows` but it does seem that there are cases for skipping before and after.  I do prefer this over the before/after boolean.  Another option would be to take in a list of row indices.  This is what pandas does.  So [0:1] skips two rows before the header (if any) and [1:2] would skip two rows after the header or [0, 2] would skip a row before and after.

@pitrou should probably take a look at this as he's got the most experience with the CSV reader.  He's out for the rest of the week so it might be next week.

**Comment by n3world** [comment]
> Thanks for submitting these PRs! I'm still brushing up on my parser knowledge but I'll take a look at the other one too.
> 
> This one looks pretty straightforward to me. It's a bit unfortunate that we have two `skip_rows` but it does seem that there are cases for skipping before and after. I do prefer this over the before/after boolean. Another option would be to take in a list of row indices. This is what pandas does. So [0:1] skips two rows before the header (if any) and [1:2] would skip two rows after the header or [0, 2] would skip a row before and after.
> 
> @pitrou should probably take a look at this as he's got the most experience with the CSV reader. He's out for the rest of the week so it might be next week.

I like the idea of replacing the current skip_rows with a list of ranges to skip. This would be a more complex but a more robust solution. I assume that skip_rows would have to be kept but deprecated. The one difficulty would be if `ParseOptions::newlines_in_values` were true then the reader would have to be aware of the rows that are parsed to have a correct row index. I can start working on this solution ignoring the newlines_in_values issue now if that is the consensus.

I might make it so any range starting with 0 is handled by the reader, as skip_row does now, and any range not starting with 0 is handled by the parser so that potentially quoted new lines are handled correctly.

There might be a way to also tackle https://issues.apache.org/jira/browse/ARROW-8527 with this change.

**Comment by westonpace** [comment]
Ah, I hadn't realized we had `newlines_in_values`.  Yes, sounds like there is a bit of complexity but you have a good grip on it.

> I assume that skip_rows would have to be kept but deprecated.

I don't think so but I'll let @pitrou weigh in.  As long as there is a good reason for it the project seems to be in a place where breaking changes are ok.  Especially since this one would be pretty trivial to understand and adapt to.

> There might be a way to also tackle https://issues.apache.org/jira/browse/ARROW-8527 with this change.

That would be good.  Eventually the dataset scanner will probably be getting a skip operation of some kind as well so that'll increase the pressure on ARROW-8527.  ARROW-12598 is also (admittedly tangentially) related since you seem to be on a roll :smile:

**Comment by n3world** [comment]
> That would be good. Eventually the dataset scanner will probably be getting a skip operation of some kind as well so that'll increase the pressure on [ARROW-8527](https://issues.apache.org/jira/browse/ARROW-8527). [ARROW-12598](https://issues.apache.org/jira/browse/ARROW-12598) is also (admittedly tangentially) related since you seem to be on a roll smile

The only tricky part about a count(*) implementation with this is that skip_rows documented that it was skipping header rows which shouldn't be counted as part of a data row count. I fee like the row count operation would have to be a little different and maybe give an indicator for on which line the actual data rows start so that the header rows before that point could be skipped.

Maybe a simpler solution would a set of two indexes: column names and first data row . While this doesn't allow arbitrary row skipping in the middle this would allow for the most common use cases, including skipping over valid rows to first desired row. With another option or operation could be used to count the number of data rows starting at first data row. The defaults would be 0, 1 for when column names are part of the csv or -1, 0 when they are not.

**Comment by westonpace** [comment]
Valid points.  In that case I wouldn't worry much about ARROW-8527 in this PR at all.  I think 8527 is really looking for "skip data rows" (even though the title says otherwise).  I'll add a comment over there.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thank you for doing this!

File: cpp/src/arrow/csv/reader_test.cc
Comment (n3world) [reply cpp/src/arrow/csv/reader_test.cc:219]: I can look into writing them in python. I just wrote the test in C++ because I implemented the feature in C++.
Right now I am actually not working on this because after the above discussions I was under the impression that the desired implementation would be to have skip_rows be a list of ranges of row indexes to skip. With that in mind I actually started to work on https://issues.apache.org/jira/browse/ARROW-12675 so that the parser could know what absolute line number it was parsing and could skip the line.

Comment (pitrou) [reply cpp/src/arrow/csv/reader_test.cc:219]: IMHO a list of ranges would be overkill. `skip_rows_after_names` seems ok to me.

Comment (n3world) [reply cpp/src/arrow/csv/reader_test.cc:219]: Actually after looking at it a bit more it doesn't have to be moved out of the reader but I don't think it can use SkipRows. SkipRows is very simple in its implementation it doesn't actually skip rows but lines in the file. I make the distinction here because a row can contain values with new lines. If any of the rows contain a quoted or escaped new line skip rows will consider that two lines and not one.

I was thinking it might be better to add add a FirstN method to Chunker to be able to get the Nth occurrence of the line endings. I was thinking this could be integrated into the BlockReader implementations to be able to skip over rows even beyond the first block. This could also solve ARROW-8527.

Comment (pitrou) [reply cpp/src/arrow/csv/reader_test.cc:219]: The fact that SkipRows is a bit dumb is a separate concern. We don't seem to have had any user complaints about it, so no need to complicate it for now, IMHO.

Comment (n3world) [reply cpp/src/arrow/csv/reader_test.cc:219]: My guess is that the reason it is simple is because of the comment that it is intended to skip corrupt rows so for that it probably has to be a bit brute force.

While adding a more csv aware skip does add some more complexity that parsing knowledge is already contained in the BoundryFinder implementations, so it already exists. Also, as a user specifying the number of rows to skip I would expect that csv rows would be skipped and not file lines.

If it sways your opinion any, I did get the implementation working that uses the BlockReaders, Chunker and BoundryFinders to skip over the lines and the parser and everything downstream are unaware. Also, it can skip lines beyond a single block, satisfying ARROW-8527.

Comment (pitrou) [reply cpp/src/arrow/csv/reader_test.cc:219]: Well, if it's not too much code and you want to do it, then why not :-)

Comment (n3world) [reply cpp/src/arrow/csv/reader_test.cc]: I updated the PR with a version which uses the BoundryFinder and can skip rows in more than one block

**Comment by n3world** [comment]
Sorry meant to delete my other branch but accidentally deleted this one.

File: cpp/src/arrow/csv/chunker_test.cc
Comment (pitrou) [inline cpp/src/arrow/csv/chunker_test.cc:86]: Can you also check the contents of `rest`?

File: cpp/src/arrow/csv/chunker_test.cc
Comment (pitrou) [inline cpp/src/arrow/csv/chunker_test.cc:297]: I don't think `ASSERT_NO_FATAL_FAILURE` is useful, but we can keep it if you want.

File: cpp/src/arrow/csv/reader.cc
Comment (pitrou) [inline cpp/src/arrow/csv/reader.cc:170]: Can you use a more precise name? Is it `skip_rows`?

File: python/pyarrow/tests/test_csv.py
Comment (pitrou) [inline python/pyarrow/tests/test_csv.py:324]: Can you perhaps add a test with both `skip_rows` and `skip_rows_after_names`?

File: cpp/src/arrow/csv/reader.cc
Comment (pitrou) [inline cpp/src/arrow/csv/reader.cc:257]: I presume this was unused?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks a lot for the update!

File: cpp/src/arrow/csv/chunker_test.cc
Comment (n3world) [reply cpp/src/arrow/csv/chunker_test.cc:297]: Without ASSERT_NO_FATAL_FAILURE the test will continue to execute even if AssertSkip fails.

With gtest assert failures in called functions do not cause the test to exit unless it is wrapped by ASSERT_NO_FATAL_FAILURE

File: cpp/src/arrow/csv/reader.cc
Comment (n3world) [reply cpp/src/arrow/csv/reader.cc:257]: Yes. I think it was in just for symmetry with BlockReader

File: cpp/src/arrow/csv/chunker_test.cc
Comment (pitrou) [reply cpp/src/arrow/csv/chunker_test.cc:297]: Ok, but how is that a problem? It would just accumulate failures instead of stopping at the first one.
Regardless, this is not important :-)

File: cpp/src/arrow/csv/chunker.cc
Comment (n3world) [reply cpp/src/arrow/csv/chunker.cc:246]: Yes because the lexer cannot be called with an empty block of data. It asserts that it is passed data and doesn't handle when `data == data_end`

File: cpp/src/arrow/csv/chunker_test.cc
Comment (n3world) [reply cpp/src/arrow/csv/chunker_test.cc:301]: It can be a problem for some tests which might cause a segfault or have bad behavior if test execution fails after an ASSERT fails, I know this is not one of those types of tests. Normally ASSERT failures halt test execution for non halting checks it is standard to use the `EXPECT_` macros

File: python/pyarrow/tests/test_csv.py
Comment (n3world) [reply python/pyarrow/tests/test_csv.py:324]: Done. Would you like more is just the one good?

File: cpp/src/arrow/util/delimiting.cc
Comment (n3world) [reply cpp/src/arrow/util/delimiting.cc]: Done. Is the comment descriptive enough?

File: cpp/src/arrow/util/delimiting.h
Comment (n3world) [reply cpp/src/arrow/util/delimiting.h]: Used two out parameters

File: python/pyarrow/tests/test_csv.py
Comment (pitrou) [reply python/pyarrow/tests/test_csv.py:324]: It's good, thank you!

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Thanks for the update @n3world ! I will merge this once CI is green.

File: cpp/src/arrow/csv/chunker.cc
Comment (n3world) [inline cpp/src/arrow/csv/chunker.cc:238]: Out of curiosity why use signed types for values which should never be negative?

Comment (pitrou) [reply cpp/src/arrow/csv/chunker.cc:238]: That's the convention we use across the Arrow codebase (for example all lengths and offsets are signed, even though they are positive). Given that C++ (and especially some compilers) can be annoying with implicit casts or mixed signed-unsigned operations, it's better not to introduce unsigned variables gratuitously.

Comment (n3world) [reply cpp/src/arrow/csv/chunker.cc:238]: Should I add asserts or checks that negative values are never passed in then because right now the code does not handle negative values?

Comment (pitrou) [reply cpp/src/arrow/csv/chunker.cc:238]: This is an internal API. We may want to do it at a higher-level, for example in the CSV reader.

Comment (n3world) [reply cpp/src/arrow/csv/chunker.cc:238]: Looking at the csv options there currently is no validation on any of those values other than asserts. It might make sense to add a options validation to the ReaderMixin as part of another PR.

Mildly related, I also noticed I probably should have made `skip_rows_after_names` an `int64_t` to match the type of the variable in the chunker. Do you want me to make that change quickly or do you want to do it?

Comment (pitrou) [reply cpp/src/arrow/csv/chunker.cc:238]: We can keep it a `int32_t` for consistency with `skip_rows`. There is zero use case for skipping more than 2**31 rows :-)

Comment (pitrou) [reply cpp/src/arrow/csv/chunker.cc:238]: You're right about options validation. Do you want to open a separate JIRA for it?

**Pull request closed without merging**
